evalml/tests/preprocessing_tests/test_balanced_classification_splitter.py

```diff
@@ -0,0 +1,343 @@
+    else:
+        assert len(X2) == 0.2 * size + 40
+        assert y2.value_counts().values[0] == 40
+
+
+def test_classification_imbalanced_multiple_multiclass():
+    X = pd.DataFrame({"a": [i for i in range(10000)]})
+    y = pd.Series([0] * 4900 + [1] * 4900 + [2] * 200)  # minority class is 2% of data
+    bcs = BalancedClassificationSampler(min_samples=201)
+    X2, y2 = bcs.fit_resample(X, y)
+    # severe imbalanace case, don't resample
+    pd.testing.assert_frame_equal(X, X2)
+    pd.testing.assert_series_equal(y, y2)
+
+    bcs = BalancedClassificationSampler()
+    X2, y2 = bcs.fit_resample(X, y)
+    assert len(X2) == 1800
+    assert all(y2.value_counts().values == [800, 800, 200])
+    assert y2.value_counts()[2] == 200
+
+    bcs = BalancedClassificationSampler(balanced_ratio=3)
+    X2, y2 = bcs.fit_resample(X, y)
+    # resample to 4:1 ratios on both 0 and 1 classes
+    assert len(X2) == 1400
+    assert all(y2.value_counts().values == [600, 600, 200])
+    assert y2.value_counts()[2] == 200
+
+
+@pytest.mark.parametrize("data_type", ['li', 'np', 'pd', 'ww'])
+def test_classification_imbalanced_data_type(data_type, make_data_type):
+    X = pd.DataFrame({"a": [i for i in range(1000)]})
+    y = pd.Series([0] * 900 + [1] * 100)
+    X = make_data_type(data_type, X)
+    y = make_data_type(data_type, y)
+
+    bcs = BalancedClassificationSampler()
+    X2, y2 = bcs.fit_resample(X, y)
+    assert len(X2) == 500
+    assert all(y2.value_counts().values == [400, 100])
+    assert y2.value_counts()[1] == 100
+
+
+def test_classification_data_frame_dtypes():
+    X = pd.DataFrame({
+        "integers": [i for i in range(1000)],
+        "strings": [f"string_{i % 3}" for i in range(1000)],
+        "text": [f"this should be text data because {i} think it's a long string. Let's hope it behaves in that way" for i in range(1000)],
+        "float": [i / 10000 for i in range(1000)],
+        "bool": [bool(i % 2) for i in range(1000)],
+        "datetime": [random.choice([2012 / 1 / 2, 2012 / 2 / 1, 2012 / 4 / 2]) for i in range(1000)]
+    })
+    y = pd.Series([0] * 900 + [1] * 100)
+    bcs = BalancedClassificationSampler()
+    X2, y2 = bcs.fit_resample(X, y)
+    assert len(X2) == 500
+    assert all(y2.value_counts().values == [400, 100])
+    assert y2.value_counts()[1] == 100
+
+    X['integers'][0] = None
+    X2, y2 = bcs.fit_resample(X, y)
+    assert len(X2) == 500
+    assert all(y2.value_counts().values == [400, 100])
+    assert y2.value_counts()[1] == 100
+
+
+def test_classification_data_drop():
+    # tests for whether or not the `max(0, counts[k] - goal_value)` code works as expected
+    X = pd.DataFrame({"a": [i for i in range(420)]})
+    y = pd.Series([0] * 90 + [1] * 100 + [2] * 120 + [3] * 40 + [4] * 70)
+    # will downsample the [2] target
+    # will try to downsample [0] and [4], but max(0, x) will prevent that
+    bcs = BalancedClassificationSampler(balanced_ratio=1, min_percentage=0.01)
```

Comment:
Mmm, yeah that's fair. My thought was that we cannot sample if we end up with less than `min_samples` values of a class. For instance, if we have 40 instances of class 1 and 100 instances of class 2, we might not want to sample the 100 instances down even if we had wanted a 1:1 ratio, whereas in a case of 40,000 vs 100,000 samples, it would be fine to sample the 100,000 to 40,000 cases since we have enough samples. As for warning or telling the user, I wasn't sure if this was something I should warn the user about or do silently. Thoughts @chukarsten @dsherry?

Reply:
I think the understanding should be that if the number of instances falls below the threshold for certain classes, as it does for "0" and "4" here, that the method will simply not downsample those classes. No error in that case.